From 74f7abdf0e567914abc12319f1e3d723e18d0d6d Mon Sep 17 00:00:00 2001
Message-Id: <74f7abdf0e567914abc12319f1e3d723e18d0d6d@dist-git>
From: =?UTF-8?q?J=C3=A1n=20Tomko?= <jtomko@redhat.com>
Date: Thu, 13 Apr 2017 12:05:48 +0200
Subject: [PATCH] qemu: do not crash on USB address with no port and invalid
 bus
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Properly error out when the user requests a port from a bus
that does not have a controller present in the domain XML.

https://bugzilla.redhat.com/show_bug.cgi?id=1441589
(cherry picked from commit b003b9781b6ae633cfe4fdf6b9620ca246fa2432)
Signed-off-by: Ján Tomko <jtomko@redhat.com>
---
 src/conf/domain_addr.c                             |  3 ++-
 .../qemuxml2argv-usb-bus-missing.xml               | 22 ++++++++++++++++++++++
 tests/qemuxml2argvtest.c                           |  3 +++
 3 files changed, 27 insertions(+), 1 deletion(-)
 create mode 100644 tests/qemuxml2argvdata/qemuxml2argv-usb-bus-missing.xml

diff --git a/src/conf/domain_addr.c b/src/conf/domain_addr.c
index 35bdc3453..8b6109199 100644
--- a/src/conf/domain_addr.c
+++ b/src/conf/domain_addr.c
@@ -2009,7 +2009,8 @@ virDomainUSBAddressAssign(virDomainUSBAddressSetPtr addrs,
 
     if (info->type == VIR_DOMAIN_DEVICE_ADDRESS_TYPE_USB) {
         VIR_DEBUG("A USB port on bus %u was requested", info->addr.usb.bus);
-        if (!addrs->buses[info->addr.usb.bus]) {
+        if (info->addr.usb.bus >= addrs->nbuses ||
+            !addrs->buses[info->addr.usb.bus]) {
             virReportError(VIR_ERR_XML_ERROR,
                            _("USB bus %u requested but no controller "
                              "with that index is present"), info->addr.usb.bus);
diff --git a/tests/qemuxml2argvdata/qemuxml2argv-usb-bus-missing.xml b/tests/qemuxml2argvdata/qemuxml2argv-usb-bus-missing.xml
new file mode 100644
index 000000000..831a4c0af
--- /dev/null
+++ b/tests/qemuxml2argvdata/qemuxml2argv-usb-bus-missing.xml
@@ -0,0 +1,22 @@
+<domain type='qemu'>
+  <name>QEMUGuest1</name>
+  <uuid>c7a5fdbd-edaf-9455-926a-d65c16db1809</uuid>
+  <memory unit='KiB'>219136</memory>
+  <currentMemory unit='KiB'>219136</currentMemory>
+  <vcpu placement='static'>1</vcpu>
+  <os>
+    <type arch='i686' machine='pc'>hvm</type>
+    <boot dev='hd'/>
+  </os>
+  <devices>
+    <emulator>/usr/bin/qemu-system-i686</emulator>
+    <controller type='usb' index='0'/>
+    <memballoon model='virtio'/>
+    <input type='mouse' bus='usb'>
+      <address type='usb' bus='0'/>
+    </input>
+    <hub type='usb'>
+      <address type='usb' bus='1'/>
+    </hub>
+  </devices>
+</domain>
diff --git a/tests/qemuxml2argvtest.c b/tests/qemuxml2argvtest.c
index cfd6dcd55..94be771d3 100644
--- a/tests/qemuxml2argvtest.c
+++ b/tests/qemuxml2argvtest.c
@@ -1349,6 +1349,9 @@ mymain(void)
     DO_TEST("usb-port-missing",
             QEMU_CAPS_CHARDEV, QEMU_CAPS_USB_HUB,
             QEMU_CAPS_NODEFCONFIG);
+    DO_TEST_PARSE_ERROR("usb-bus-missing",
+                        QEMU_CAPS_CHARDEV, QEMU_CAPS_USB_HUB,
+                        QEMU_CAPS_NODEFCONFIG);
     DO_TEST("usb-ports",
             QEMU_CAPS_CHARDEV, QEMU_CAPS_USB_HUB,
             QEMU_CAPS_NODEFCONFIG);
-- 
2.12.2

